### PR TITLE
Use the output hierarchy to check for overlapping outputs in execution plan

### PR DESCRIPTION
**Pull request by wolfs**
This addresses some more of the performance regression here: https://github.com/gradle/gradle/issues/16048

We now use the output hierarchy information to check whether a task can be scheduled with the currently running nodes. That makes things a little bit faster again (another 2-3%). So together with https://github.com/gradle/gradle/pull/16178 we should be back to the performance of 6.8.2.

### CODE REVIEW

**Comment by wolfs** [comment]
@bot-gradle test QFL

File: subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java [hunk 1]
```diff
@@ -99,14 +100,16 @@ public DefaultExecutionPlan(
         TaskDependencyResolver dependencyResolver,
         NodeValidator nodeValidator,
         ExecutionNodeAccessHierarchy outputHierarchy,
-        ExecutionNodeAccessHierarchy inputHierarchy
+        ExecutionNodeAccessHierarchy inputHierarchy,
+        ExecutionNodeAccessHierarchy destroyableHierarchy
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java:104]: Nit: would be nice to get these in some logical order.

**Comment by wolfs** [comment]
Here is an non-abi change benchmark on the Uber project: https://builds.gradle.org/repository/download/Gradle_Master_Util_Performance_AdHocPerformanceScenarioLinux/41514532:id/results/performance/build/test-results-uberMobileAppPerformanceAdHocTest.zip!/non_abi_change_with_configuration_caching/benchmark.html

image

This change makes things a little bit slower and removes the extra code to check for overlay between outputs.

**Review by lptr — COMMENTED** [review]
I'm not sure I understand: are we making things faster this way, or are we making the code more readable only?

File: subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java (hunk 1 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java:104]: Logical meaning input/output/destroys?

Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java:104]: That's a good one.

Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java:104]: Alphabetical would also work I think, but I like yours better.

**Comment by wolfs** [comment]
> I'm not sure I understand: are we making things faster this way, or are we making the code more readable only?

We are making things a little bit faster (around 2%) and we are replacing an n * m algorithm (checking every output path with every other output path) by an n log m algorithm (looking up the affected nodes in a tree for every output path).

If we would only have the measured improvement of 2%, then I'd say the change is not worth it, though since we are making the code better to understand by re-using data we already have, I'd go for this.

File: subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java (hunk 1 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java:104]: Done.

**Pull request merged**